tests/dualtor/test_tunnel_memory_leak.py

```diff
@@ -0,0 +1,183 @@
+                    .format(mem_percent, swss_mem_percent))
+        return True
+    return False
+
+def test_tunnel_memory_leak(
+    toggle_all_simulator_ports_to_upper_tor,
+    upper_tor_host, lower_tor_host, ptfhost, 
+    ptfadapter, conn_graph_facts, tbinfo, vmhost
+):
+    """
+    Test if there is memory leak for service tunnel_packet_handler.
+    Send ip packets from standby TOR T1 to Server, standby TOR will
+    forward the packets to active TOR with tunnel, active TOR will
+    decapsulate the IPinIP packets, but there is no neighbor for destination
+    as we remove neighbor before test, tunnel_packet_handler will be
+    triggered and neighbor will be added. Server will receive the packets.
+    Check if memory usage is increased after tunnel_packet_handler's
+    operation. Since tunnel_packet_handler is only triggered by the
+    first packet, loop the process for all severs to trigger it as much
+    as possible.
+    """
+    @contextlib.contextmanager
+    def prepare_services(ptfhost):
+        """
+        Temporarily stop garp service, start arp and icmp service.
+        It has to stop garp_service for triggering tunnel_packet_handler.
+        It has to start arp and icmp service for receiving packets at server side. 
+        """
+        ptfhost.shell("supervisorctl stop garp_service")
+        ptfhost.shell("supervisorctl start arp_responder")
+        ptfhost.shell("supervisorctl start icmp_responder")
+        yield
+        ptfhost.shell("supervisorctl start garp_service")
+        ptfhost.shell("supervisorctl stop arp_responder")
+        ptfhost.shell("supervisorctl stop icmp_responder")
+	
+    @contextlib.contextmanager
+    def remove_neighbor(duthost, server_ip):
+        """
+        Remove ip neighbor before test for triggering tunnel_packet_handler,
+        restore it after test
+        """
+        flush_neighbor_ct = flush_neighbor(duthost, server_ip, True)
+
+        with flush_neighbor_ct:
+            command = "ip neighbor show %s" % server_ip
+            output = [_.strip() for _ in duthost.shell(command)["stdout_lines"]]
+            pytest_assert(not output, "server ip {} isn't flushed in neighbor table.".format(server_ip))
+            yield
+
+    pytest_assert(is_tunnel_packet_handler_running(upper_tor_host), 
+                "tunnel_packet_handler is not running in SWSS conainter.")
+
+    ptf_t1_intf = random.choice(get_t1_ptf_ports(lower_tor_host, tbinfo))
+
+    all_servers_ips = mux_cable_server_ip(upper_tor_host)
+    # Get the original memeory percent before test.
+    check_memory_leak(upper_tor_host)
+    with prepare_services(ptfhost):
+        for iface, server_ips in all_servers_ips.items():
+            server_ipv4 = server_ips["server_ipv4"].split("/")[0]
+            logging.info("Select DUT interface {} and server IP {} to test.".format(iface, server_ipv4))
+
+            pkt, exp_pkt = build_packet_to_server(lower_tor_host, ptfadapter, server_ipv4)
```

Comment:
I think it's better to directly build an IPinIP packet and send it from the PTF straight to the active ToR, it's not necessary to involved the peer ToR for this test.

Reply:
Also, please make sure to set the chosen DUT to active prior to sending the packets

Reply:
As I mentioned early, I test the whole packet process on dualtor topology, Standby Tor T1-> Standby Tor -> Active Tor ->Server.
This is why I don't build an IPinIP packet.

Reply:
Makes sense, misunderstanding on my part.